Advertisement creation

YegorPopinako opened the pull request:
added api for advertisement creation
added unit tests

YegorPopinako committed:
feat(api): add APi for advert creation

YegorPopinako committed:
merged from 'main' to advertisement-creation

YegorPopinako committed:
refactor(api): fullfilled controller and service methods

YegorPopinako committed:
merged from 'main' to advertisement-creation

YegorPopinako committed:
merged from 'main' to advertisement-creation

YegorPopinako committed:
feat(dto): add dto for advert creation

YegorPopinako committed:
feat(exc): implement UserNotFoundException for missing user handling

YegorPopinako committed:
refactor(api): improved api for advert creation

YegorPopinako committed:
merged from 'main' to advertisement-creation

YegorPopinako committed:
test(unit): add tests for AdvertService

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/JruGotoM5Application.java:
```diff
@@ -1,15 +1,13 @@
     }
-}
 
+}
```
what sense of this change?

YegorPopinako replied:
cant remember that i changed it, maybe by an accident

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/controller/AdvertController.java:
```diff
@@ -1,29 +1,31 @@
         return advertService.getAdvertsByDateRange(from, to);
     }
+
+    @PostMapping("/new")
```
don't use verbs in url

YegorPopinako replied:
'new' is not a verb, it can be an adjective, adverb, or a noun, but never a verb

Maksym82564 commented on src/test/java/com/codegym/jrugotom5/service/AdvertServiceTest.java:
```diff
@@ -80,4 +86,41 @@ void testGetAdvertsByDateRange_InvalidDateRange() {
+
+        advertService.createAdvert(advertCreateDTO);
+
+        verify(advertRepository, times(1)).save(any(Advert.class));
```
https://github.com/Volodymyr95/goto-m5f/pull/5#discussion_r1803620359

Maksym82564 commented on src/main/java/com/codegym/jrugotom5/dto/AdvertCreateDTO.java:
```diff
@@ -0,0 +1,30 @@
+    private String description;
+
+    @NotNull
+    private User createdBy;
```
entity object does not belong in dto. you receive data through enpoint, then it serializes to your dto, entity objects are not serializable by default and they are not meant for such use. 
user id should be there instead. take this id -> userRepository.findById() -> insert user obj into advert obj -> save advert obj.
@Volodymyr95 am i right?

YegorPopinako replied:
I'll wait for mentor to check this issue before fixing

Maksym82564 replied:
did you try to call your endpoint with postman?

YegorPopinako replied:
no but as far as im concerned the field createdBy can be pasted like this
    "createdBy": {
        "id": 1
    }
however i read about this issue and yeah, you are right, i will fix it ASAP, thanks)

Maksym82564 commented on src/main/java/com/codegym/jrugotom5/service/AdvertService.java:
```diff
@@ -30,4 +36,15 @@ public List<AdvertDTO> getAdvertsByDateRange(LocalDate from, LocalDate to) {
     }
+
+    @Transactional
+    public void createAdvert(AdvertCreateDTO advert) {
```
name this param "dto" or "advertDto" instead of "advert" for better clarity

YegorPopinako committed:
merged from 'main' to advertisement-creation

YegorPopinako committed:
refactor: minor code fixes

YegorPopinako committed:
refactor(dto): remove User entity from dto, use authorId instead

YegorPopinako committed:
fix(service): make service set the User to Advert entity

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/controller/AdvertController.java:
```diff
@@ -16,7 +19,8 @@
 @RequestMapping("/api")
```
/api/adeverts

Maksym82564 replied:
> /api/adeverts

did this in #8, should be in main by now

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/controller/AdvertController.java:
```diff
@@ -16,7 +19,8 @@
+
+   private final AdvertService advertService;
 
     @GetMapping(path = "/adverts")
```
/ GET

YegorPopinako replied:
could you please provide more detailed info on this comment?

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/controller/AdvertController.java:
```diff
@@ -27,4 +31,9 @@ public List<AdvertBasicInfoDTO> getAllAdverts() {
         return advertService.getAdvertsByDateRange(from, to);
     }
+
+    @PostMapping("/new")
```
/ POST

YegorPopinako replied:
could you please provide more detailed info on this comment?

Volodymyr95 replied:
API url : /api/adverts/ POST

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/controller/AdvertController.java:
```diff
@@ -27,4 +31,9 @@ public List<AdvertBasicInfoDTO> getAllAdverts() {
     }
+
+    @PostMapping("/new")
+    public void createAdvert(@RequestBody @Valid AdvertCreateDTO advert) {
```
why return void?

YegorPopinako replied:
changed to ResponseEntity

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/controller/AdvertController.java:
```diff
@@ -27,4 +31,9 @@ public List<AdvertBasicInfoDTO> getAllAdverts() {
+
+    @PostMapping("/new")
+    public void createAdvert(@RequestBody @Valid AdvertCreateDTO advert) {
+        advertService.createAdvert(advert);
```
return 201 status code

YegorPopinako committed:
merged from 'main'

YegorPopinako committed:
refactor(api): make advert create endpoint give body and response status

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/exception/GlobalExceptionHandler.java:
```diff
@@ -11,4 +11,9 @@ public class GlobalExceptionHandler {
+
+    @ExceptionHandler(UserNotFoundException.class)
+    public ResponseEntity<String> handleUserNotFound(UserNotFoundException ex) {
+        return new ResponseEntity<>(ex.getMessage(), HttpStatus.NOT_FOUND);
```
missed logger

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/service/AdvertService.java:
```diff
@@ -44,4 +49,19 @@ public List<AdvertBasicInfoDTO> getAllAdverts() {
+        }
+        Advert advertEntity = modelMapper.map(advertCreateDTO, Advert.class);
+        advertEntity.setCreatedDate(LocalDate.now());
+        advertEntity.setEndDate(LocalDate.now().plusDays(30));
```
30 -make it constant, it's magic number

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/service/AdvertService.java:
```diff
@@ -44,4 +49,19 @@ public List<AdvertBasicInfoDTO> getAllAdverts() {
+        advertEntity.setEndDate(LocalDate.now().plusDays(30));
+        advertEntity.setCreatedBy(userService.getUserById(id));
+        advertRepository.save(advertEntity);
+        return modelMapper.map(advertEntity, AdvertFullInfoDTO.class);
```
return modelMapper.map(advertRepository.save(advertEntity), AdvertFullInfoDTO.class);

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/service/UserService.java:
```diff
@@ -22,4 +24,13 @@ public List<UserBasicInfoDTO> getAllUsers() {
                 .toList();
     }
+
+    public User getUserById(Long id) {
```
why u return Entity?

YegorPopinako replied:
because the User's entity is required to be set to the Advert entity, the retrieved User entity is used only for inner logic

Volodymyr95 replied:
u don't have repository or mapper?

YegorPopinako committed:
fix(api): minor code fixes

YegorPopinako committed:
chore(props): stop tracking local properties

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/controller/AdvertController.java:
```diff
@@ -27,4 +29,10 @@ public List<AdvertBasicInfoDTO> getAllAdverts() {
         return advertService.getAdvertsByDateRange(from, to);
     }
+
+    @PostMapping("/POST")
```
facepalm

Volodymyr95 replied:
what is a POST in REST and HTTP?

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/exception/GlobalExceptionHandler.java:
```diff
@@ -1,14 +1,22 @@
+
+    @ExceptionHandler(UserNotFoundException.class)
+    public ResponseEntity<String> handleUserNotFound(UserNotFoundException ex) {
+        log.error("Exception occurred: {}", ex.getMessage());
```
is just message informative?

YegorPopinako replied:
added a timestamp and a status code, if there is any more info an exception could give could you please inform me

Elder-HD replied:
i suppose mentor means that you put not enough information in your LOG message (not information for client). And it relates to both of your exception handlers. It will be hard to debug application when you will have just only "Message of your exception"
see:
https://github.com/Volodymyr95/goto-m5f/pull/16#discussion_r1812546592
https://github.com/Volodymyr95/goto-m5f/pull/13#discussion_r1812556175

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/service/AdvertService.java:
```diff
@@ -19,8 +22,11 @@
 @Slf4j
 public class AdvertService {
+
+    private final Integer DAYS_TO_END = 30;
```
https://www.javatpoint.com/java-constant

YegorPopinako committed:
chore(api): code refactoring

YegorPopinako committed:
chore(exception and test): code refactoring

Maksym82564 commented on src/main/java/com/codegym/jrugotom5/service/AdvertService.java:
```diff
@@ -41,7 +50,22 @@ public List<AdvertFullInfoDTO> getAdvertsByDateRange(LocalDate from, LocalDate t
+    public AdvertFullInfoDTO createAdvert(AdvertCreateDTO advertCreateDTO) {
+        Long id = advertCreateDTO.getUserCreatorId();
+        if (!userService.userExistsById(id)) {
+            log.error("Could not find User with id {}", id);
```
You log this error 2 times: here and in GlobalExceptionHandler. Maybe remove logging from here?

Elder-HD replied:
I agree with Maksym

YegorPopinako committed:
feat(exception): add validation exceptions handling

Elder-HD commented on src/main/java/com/codegym/jrugotom5/service/AdvertService.java:
```diff
@@ -41,7 +50,22 @@ public List<AdvertFullInfoDTO> getAdvertsByDateRange(LocalDate from, LocalDate t
+            log.error("Could not find User with id {}", id);
+            throw new UserNotFoundException("User with id " + id + " does not exist.");
+        }
+        Advert advertEntity = modelMapper.map(advertCreateDTO, Advert.class);
```
Your mapper conversion doesn't work! Try to debug and check "ID" value of advertEntity.
On this step it must have 'null' value.
It seems `modelMapper` converts `userCreatorId` (from DTO) to `id` of advert (to entity).

Also i suppose you should set `userCreatorId` for your resulting DTO `AdvertFullInfoDTO` manually after mapping.

Elder-HD commented on src/main/java/com/codegym/jrugotom5/dto/AdvertCreateDTO.java:
Your validation doesn't work. 
i removed the title and used length of it lower than 5 symbols but there wasn't any validation
Did you even check it?

Elder-HD commented on src/main/java/com/codegym/jrugotom5/controller/AdvertController.java:
```diff
@@ -27,4 +29,10 @@ public List<AdvertBasicInfoDTO> getAllAdverts() {
+    @PostMapping
+    public ResponseEntity<AdvertFullInfoDTO> createAdvert(@RequestBody @Valid AdvertCreateDTO advertCreateDTO) {
+        AdvertFullInfoDTO advertFullInfoDTO = advertService.createAdvert(advertCreateDTO);
+        return new ResponseEntity<>(advertFullInfoDTO, HttpStatus.CREATED);
```
I suppose it's better to use Method annotation (it's more clear) instead of "ResponceEntity"

Elder-HD commented on src/main/resources/application-local.properties:
You deleted this file! So after merge of this PR all student will lose this file too. 
Please check all your commits and don't delete any COMMON files (not related to your task)

Elder-HD reviewed (changes requested):
Your method and your validation doen't work correctly - see comments
Try to test it with non empty DB (at least 5 users and 10 adverts)

YegorPopinako committed:
merged from main

YegorPopinako committed:
chore(api): code fixes

YegorPopinako committed:
merged from main

YegorPopinako committed:
style(repository): style formatting

YegorPopinako committed:
test(unit): add test for required test coverage

YegorPopinako committed:
test(unit): add test for required test coverage

YegorPopinako commented:
вавава

YegorPopinako committed:
refactor(service): extract method for finding user

YegorPopinako committed:
test(unit): add unit tests for fulfilling test coverage requirements

YegorPopinako commented:
After last merge the test coverage verification did not pass
Covered ratio was 0.5 with 0.6 required
According to the Jacoco test report the method **getAdvertsByDateRange** was not fully covered so i added additional test for it called **testGetAdvertsByDateRange_FromDateEqualToDate**
Also refactored createAdvert method by exctracting logic for getting user from DB into a separate method 
So now test coverage is 62%

P.S. @Elder-HD my thanks for reviewing my code, i checked and tested the updated logic and everything has to work as it supposed to

Elder-HD commented on src/main/java/com/codegym/jrugotom5/service/AdvertService.java:
```diff
@@ -49,4 +59,33 @@ public List<AdvertBasicInfoDTO> getByTitleContains(String phrase) {
+        return advertFullInfoDTO;
+    }
+
+    public User findUserByIdOrThrow(Long id) {
```
i think this method is redurant.
Optinal (which is returned by method `findById(id)`) already has `orElseThrow(Supplier)` method that do same logic (when optional value null or empty).

But even more. You will never throw UserNotFoundException on this step as you have the same check in row #66. So you can remove check-throw from row #66 or #76

Volodymyr95 commented on src/main/java/com/codegym/jrugotom5/controller/AdvertController.java:
```diff
@@ -32,4 +33,10 @@ public List<AdvertFullInfoDTO> getAdverts(@RequestParam LocalDate from, LocalDat
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public AdvertFullInfoDTO createAdvert(@RequestBody @Valid AdvertCreateDTO advertCreateDTO) {
```
@Validated

YegorPopinako committed:
merged from main

YegorPopinako committed:
chore(api): minor code fixes

YegorPopinako committed:
test(integration): add integration test for create advert api

Elder-HD commented on src/main/java/com/codegym/jrugotom5/service/AdvertService.java:
```diff
@@ -64,4 +73,27 @@ public List<AdvertBasicInfoDTO> getByCategory(String category) {
+    @Transactional
+    public AdvertFullInfoDTO createAdvert(AdvertCreateDTO advertCreateDTO) {
+        Long id = advertCreateDTO.getUserCreatorId();
+        if (!userService.userExistsById(id)) {
+            throw new UserNotFoundException("User with id " + id + " does not exist.");
+        }
```
Yeap, i would move `findById` (row #91-92) here instead of `userExistsById` (remove if-check)

Maksym82564 replied:
This logic, along with the userExistsById method, is redundant since `User user = userRepository.findById(id).orElseThrow()` already throws exception if user does not exist.

YegorPopinako replied:
yeah, for sure, thanks

YegorPopinako committed:
chore(service): minor code fixes

YegorPopinako committed:
test(unit): rewrite test according to new service logic

YegorPopinako committed:
docs(api): document response codes for advert creation

YegorPopinako committed:
chore: merged from main
